File: mechasnippet/inserter.py
"""Inserción del snippet en la app activa.

Antes de tocar nada verifica el permiso de Accesibilidad: si falta, NO borra el
disparador (para no comerse texto del usuario) ni pega. Preserva TODO el
portapapeles del usuario (no solo texto: imágenes, archivos, RTF, etc.), marca
el snippet como efímero para que los gestores de historial no lo capturen, y
restaura en el hilo principal solo si nadie más tocó el portapapeles entremedio.
"""

import sys
import traceback

import Quartz
from AppKit import NSPasteboard, NSPasteboardItem, NSPasteboardTypeString
from ApplicationServices import AXIsProcessTrusted
from PyObjCTools import AppHelper

KEY_DELETE = 51  # tecla Borrar (backspace)
KEY_V = 9        # tecla "v"
RESTORE_DELAY = 0.5  # segundos antes de devolver el portapapeles original

# Tipos que avisan a los gestores de portapapeles (Raycast, Maccy, Paste) que
# este contenido es sensible/transitorio y NO deben guardarlo en su historial.
CONCEALED_TYPE = "org.nspasteboard.ConcealedType"
TRANSIENT_TYPE = "org.nspasteboard.TransientType"


def _post_key(keycode, flags=0):
    source = Quartz.CGEventSourceCreate(Quartz.kCGEventSourceStateHIDSystemState)
    down = Quartz.CGEventCreateKeyboardEvent(source, keycode, True)
    up = Quartz.CGEventCreateKeyboardEvent(source, keycode, False)
    if flags:
        Quartz.CGEventSetFlags(down, flags)
        Quartz.CGEventSetFlags(up, flags)
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, down)
    Quartz.CGEventPost(Quartz.kCGHIDEventTap, up)


def _snapshot(pasteboard):
    """Copia TODOS los tipos de TODOS los items del portapapeles."""
    snapshot = []
    for item in (pasteboard.pasteboardItems() or []):
        data = {}
        for t in item.types():
            payload = item.dataForType_(t)
            if payload is not None:
                data[t] = payload
        if data:
            snapshot.append(data)
    return snapshot


def _restore(pasteboard, snapshot, expected_change):
    # Si alguien más modificó el portapapeles desde que pegamos, no lo pisamos.
    if pasteboard.changeCount() != expected_change:
        return
    pasteboard.clearContents()
    items = []
    for data in snapshot:
        item = NSPasteboardItem.alloc().init()
        for t, payload in data.items():
            item.setData_forType_(payload, t)
        items.append(item)
    if items:
        pasteboard.writeObjects_(items)


def insert_snippet(content, backspaces=2):
    """Borra el disparador y pega `content` en la app que tenga el foco."""
    try:
        if not AXIsProcessTrusted():
            # Sin Accesibilidad, CGEventPost no tiene efecto. No borramos el
            # disparador ni tocamos el portapapeles, para no dejar texto comido.
            sys.stderr.write(
                "[mecha] Sin permiso de Accesibilidad: no se pega ni se borra el disparador.\n"
            )
            sys.stderr.flush()
            return

        sys.stderr.write(
            "[mecha] insert_snippet: borrando %d, pegando %d chars\n"
            % (backspaces, len(content))
        )
        sys.stderr.flush()

        pasteboard = NSPasteboard.generalPasteboard()
        saved = _snapshot(pasteboard)

        for _ in range(backspaces):
            _post_key(KEY_DELETE)

        # Escribir el snippet marcado como efímero/oculto.
        pasteboard.clearContents()
        item = NSPasteboardItem.alloc().init()
        item.setString_forType_(content, NSPasteboardTypeString)
        item.setString_forType_("", CONCEALED_TYPE)
        item.setString_forType_("", TRANSIENT_TYPE)
        pasteboard.writeObjects_([item])
        our_change = pasteboard.changeCount()

        _post_key(KEY_V, Quartz.kCGEventFlagMaskCommand)

        # Restaurar en el hilo principal y solo si nadie más tocó el portapapeles.
        AppHelper.callLater(RESTORE_DELAY, _restore, pasteboard, saved, our_change)
    except Exception:
        sys.stderr.write("[mecha] ERROR en insert_snippet:\n")
        traceback.print_exc()
        sys.stderr.flush()

File: Sources/MechaSnippet/Hotkey/Inserter.swift
import AppKit
import CoreGraphics
import ApplicationServices

/// Pega un snippet en la app activa: borra el disparador, escribe el snippet en
/// el portapapeles (marcado como efímero), envía Cmd+V y restaura el
/// portapapeles original. Sin Accesibilidad no hace nada (para no comer texto).
enum Inserter {
    private static let keyDelete: CGKeyCode = 51
    private static let keyV: CGKeyCode = 9
    private static let concealed = "org.nspasteboard.ConcealedType"
    private static let transient = "org.nspasteboard.TransientType"
    private static let restoreDelay = 0.5

    static func insert(_ content: String, backspaces: Int = 2) {
        guard AXIsProcessTrusted() else {
            NSLog("[mecha] sin Accesibilidad: no se pega ni se borra el disparador")
            return
        }
        let pb = NSPasteboard.general
        let saved = snapshot(pb)

        for _ in 0..<backspaces { postKey(keyDelete) }

        pb.clearContents()
        let item = NSPasteboardItem()
        item.setString(content, forType: .string)
        item.setString("", forType: NSPasteboard.PasteboardType(concealed))
        item.setString("", forType: NSPasteboard.PasteboardType(transient))
        pb.writeObjects([item])
        let ourChange = pb.changeCount

        postKey(keyV, flags: .maskCommand)

        DispatchQueue.main.asyncAfter(deadline: .now() + restoreDelay) {
            restore(pb, saved, expected: ourChange)
        }
    }

    private static func postKey(_ code: CGKeyCode, flags: CGEventFlags = []) {
        let src = CGEventSource(stateID: .hidSystemState)
        let down = CGEvent(keyboardEventSource: src, virtualKey: code, keyDown: true)
        let up = CGEvent(keyboardEventSource: src, virtualKey: code, keyDown: false)
        down?.flags = flags
        up?.flags = flags
        down?.post(tap: .cghidEventTap)
        up?.post(tap: .cghidEventTap)
    }

    private static func snapshot(_ pb: NSPasteboard) -> [[NSPasteboard.PasteboardType: Data]] {
        (pb.pasteboardItems ?? []).map { item in
            var d: [NSPasteboard.PasteboardType: Data] = [:]
            for t in item.types { if let data = item.data(forType: t) { d[t] = data } }
            return d
        }
    }

    private static func restore(
        _ pb: NSPasteboard,
        _ snap: [[NSPasteboard.PasteboardType: Data]],
        expected: Int
    ) {
        guard pb.changeCount == expected else { return }  // alguien más lo tocó
        pb.clearContents()
        let items = snap.map { d -> NSPasteboardItem in
            let it = NSPasteboardItem()
            for (t, data) in d { it.setData(data, forType: t) }
            return it
        }
        if !items.isEmpty { pb.writeObjects(items) }
    }
}
